### PR TITLE
Syncing improvements

### PR DESCRIPTION
**Pull request by mfrankovi**
This PR is missing a more consistent and comprehensive handling of malicious nodes which I believe is another large task that needs to be done. As part of this PR I have left a lot of TODO: for places where such malicious node handling might be needed. 

Changes are:

**PBFT syncing**

Pbft manager no longer controls the syncing process. Status messages are used for restarting pbft syncing if needed. Previously a lot of time restartSyncing was called in cases where there was no need for it. Pbft syncing is started only if:
1. There is a node at least 2 pbft periods ahead
2. There is a node 1 pbft period ahead for the same period for two consecutive status messages  

**DAG syncing**

DAG syncing is refactored and simplified:
Each connected peer will request all the missing DAG blocks level from a connected peer in only one of these two cases:
1. On connecting if the pbft periods of both nodes are matching
2. On receiving a first dag block with mising tip/pivot if node is pbft synced

This request can only be performed once. After this is performed it is expected that nodes are DAG synced. If after this sync, another block with missing tip/pivot is received a node is disconnected.

Main difference to what we had before is that there is no global dag_syncing flag anymore. We are now ensuring dag syncing state between each connected peer. As long as this is true DAG gossiping should work without any missing tip/pivot errors.

### CODE REVIEW

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 1]
```diff
@@ -611,9 +611,9 @@ void DagManager::recoverDag() {
   trx_mgr_->recoverNonfinalizedTransactions();
 }
 
-const std::map<uint64_t, std::vector<blk_hash_t>> DagManager::getNonFinalizedBlocks() const {
+const std::pair<uint64_t, std::map<uint64_t, std::vector<blk_hash_t>>> DagManager::getNonFinalizedBlocks() const {
   SharedLock lock(mutex_);
-  return non_finalized_blks_;
+  return {period_, non_finalized_blks_};
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/src/dag/dag.cpp:616]: Non finalized DAG blocks should have no period yet. What is the period number?

File: libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp [hunk 2]
```diff
@@ -46,23 +41,28 @@ void DagBlockPacketHandler::process(const PacketData &packet_data, const std::sh
...
-      status.second.insert(hash);
-      requestDagBlocks(packet_data.from_node_id_, status.second, DagSyncRequestType::MissingHashes);
-      return;
+      if (peer->peer_dag_syncing) {
+        LOG(log_dg_) << "Ignore new dag block " << hash.abridged() << ", dag syncing is on";
+      } else {
+        if (peer->peer_dag_synced) {
+          LOG(log_er_) << "DagBlock" << block.getHash() << " has missing pivot or/and tips " << status.second
+                       << " . Peer " << packet_data.from_node_id_.abridged() << " will be disconnected.";
+          disconnect(peer->getId(), dev::p2p::UserReason);
```
Comment (gaoqi8649) [inline libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp:56]: I think here is dangerous. DAG is growing constantly, that's very hard to determined peer DAG has synced or not

File: libraries/core_libs/network/src/tarcap/packets_handlers/pbft_block_packet_handler.cpp [hunk 3]
```diff
@@ -28,6 +33,19 @@ void PbftBlockPacketHandler::process(const PacketData &packet_data, const std::s
     peer->pbft_chain_size_ = peer_pbft_chain_size;
   }
 
+  // Ignore any pbft block that is missing dag anchor
```
Comment (gaoqi8649) [inline libraries/core_libs/network/src/tarcap/packets_handlers/pbft_block_packet_handler.cpp:36]: If this is a good PBFT proposal block, just DAG has not synced yet. We should not drop the block

**Comment by gaoqi8649** [comment]
After carefully thinking of this, I have several questions:

// PBFT anchor gossiping
PBFT is using votes to determine voting value, not by proposal blocks. If a node receive votes but drop the PBFT proposal block(because of missing DAG blocks). That will still vote on the value until timeout giving up on the value. 
If DAG blocks proposal in a high rate, that will cause there is no enough(2t+1) nodes have the same DAG in time. And that will cause continue to propose PBFT blocks and never finalize blocks. 
Since the main bug is DAG blocks syncing problem for missing DAG blocks issue. The change just hidden the bug in PBFT side, not fix the bug. If missing DAG blocks issue still there, PBFT still cannot finalize blocks, and instead of proposing new blocks constantly. If the missing DAG blocks bug get fixed, we don’t need change anything here without consider malicious players.

So I think this is only working for protect from malicious players. I think should be no help on DAG syncing or PBFT finalization. 
And how to check the PBFT block sender is malicious players and not because of DAG has not synced yet?

// PBFT syncing. 
This is OK for Testnet, but I think that’s not good for Mainnet. Actually I have a completely opposite view. I think we should avoid sending any syncing request by state messages because state messages cannot verify, malicious players could send out state message with any number.

// DAG syncing
DAG is growing constantly. How to determine a node’s DAG has synced or not?

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 1 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:616]: This is the current period we are in. Basically the same values as current pbft chain size

File: libraries/core_libs/network/src/tarcap/packets_handlers/pbft_block_packet_handler.cpp (hunk 3 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/pbft_block_packet_handler.cpp]: I have modified this

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp [hunk 4]
```diff
@@ -105,49 +106,52 @@ void StatusPacketHandler::process(const PacketData& packet_data, const std::shar
...
-    selected_peer->pbft_round_ = (*it++).toPositiveInt64();
+    selected_peer->pbft_round_ = (*it++).toInt<uint64_t>();
     selected_peer->pbft_previous_round_next_votes_size_ = (*it++).toInt<unsigned>();
 
+    // TODO: Address malicious status
+    if (!syncing_state_->is_pbft_syncing()) {
+      if (pbft_synced_period < selected_peer->pbft_chain_size_) {
+        LOG(log_nf_) << "Restart PBFT chain syncing. Own synced PBFT at period " << pbft_synced_period
+                     << ", peer PBFT chain size " << selected_peer->pbft_chain_size_;
+        if (pbft_synced_period < selected_peer->pbft_chain_size_ + 1) {
```
Comment (gaoqi8649) [inline libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:120]: Dose this should be "(pbft_synced_period + 1) < selected_peer->pbft_chain_size_"?

File: libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp (hunk 2 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp:56]: This flag indicates that we at least once did pull all the non-finalized dag blocks from the node sending us this message. If this has already happened and if there is a missing tip/pivot node is disconnected since this should not be possible with honest node

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 4 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp]: Yes. Good catch. I fixed it

File: libraries/core_libs/network/include/network/tarcap/packets_handlers/get_dag_sync_packet_handler.hpp [hunk 5]
```diff
@@ -20,7 +20,8 @@ class GetDagSyncPacketHandler : public PacketHandler {
 
   virtual ~GetDagSyncPacketHandler() = default;
 
-  void sendBlocks(dev::p2p::NodeID const& peer_id, std::vector<std::shared_ptr<DagBlock>> blocks);
+  void sendBlocks(dev::p2p::NodeID const& peer_id, std::vector<std::shared_ptr<DagBlock>> blocks,
```
Comment (JakubFornadel) [inline libraries/core_libs/network/include/network/tarcap/packets_handlers/get_dag_sync_packet_handler.hpp:23]: std::vector<std::shared_ptr<DagBlock>> should be probably `const ref` - a lot of data has to be copied here

File: libraries/core_libs/network/include/network/tarcap/packets_handlers/pbft_block_packet_handler.hpp [hunk 6]
```diff
@@ -6,10 +6,13 @@ namespace taraxa {
 class PbftBlock;
 class PbftChain;
 class PbftManager;
+class DagBlockManager;
 }  // namespace taraxa
 
 namespace taraxa::network::tarcap {
 
+class SyncingState;
```
Comment (JakubFornadel) [inline libraries/core_libs/network/include/network/tarcap/packets_handlers/pbft_block_packet_handler.hpp:14]: why are these forwards declarations in this header when they are not used here ?

File: libraries/core_libs/network/include/network/tarcap/taraxa_capability.hpp [hunk 7]
```diff
@@ -83,7 +83,8 @@ class TaraxaCapability : public dev::p2p::CapabilityFace {
 
   // METHODS USED IN TESTS ONLY
   void sendBlock(dev::p2p::NodeID const &id, DagBlock const &blk);
-  void sendBlocks(dev::p2p::NodeID const &id, std::vector<std::shared_ptr<DagBlock>> blocks);
+  void sendBlocks(dev::p2p::NodeID const &id, std::vector<std::shared_ptr<DagBlock>> blocks, uint64_t request_period,
```
Comment (JakubFornadel) [inline libraries/core_libs/network/include/network/tarcap/taraxa_capability.hpp:86]: also here `const ref` for blocks

File: libraries/core_libs/network/include/network/tarcap/taraxa_peer.hpp [hunk 8]
```diff
@@ -70,6 +70,12 @@ class TaraxaPeer : public boost::noncopyable {
   std::atomic<uint64_t> pbft_chain_size_ = 0;
   std::atomic<uint64_t> pbft_round_ = 1;
   std::atomic<size_t> pbft_previous_round_next_votes_size_ = 0;
+  std::atomic<uint64_t> last_status_pbft_chain_size_ = 0;
+  std::atomic_bool peer_dag_synced = false;
+  std::atomic_bool peer_dag_syncing = false;
```
Comment (JakubFornadel) [inline libraries/core_libs/network/include/network/tarcap/taraxa_peer.hpp:75]: missing underscore at the end

File: libraries/core_libs/network/include/network/tarcap/threadpool/packets_blocking_mask.hpp [hunk 9]
```diff
@@ -6,6 +6,8 @@
 #include <map>
 #include <shared_mutex>
 
+#include "config/config.hpp"
+#include "logger/logger.hpp"
```
Comment (JakubFornadel) [inline libraries/core_libs/network/include/network/tarcap/threadpool/packets_blocking_mask.hpp:10]: why were these included added when not used here ?

File: libraries/core_libs/network/src/network.cpp [hunk 10]
```diff
@@ -128,9 +128,10 @@ void Network::sendBlock(dev::p2p::NodeID const &id, DagBlock const &blk) {
   LOG(log_dg_) << "Sent Block:" << blk.getHash().toString();
 }
 
-void Network::sendBlocks(dev::p2p::NodeID const &id, std::vector<std::shared_ptr<DagBlock>> blocks) {
+void Network::sendBlocks(dev::p2p::NodeID const &id, std::vector<std::shared_ptr<DagBlock>> blocks,
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/network.cpp:131]: hmm I see that we have a lot of chaos in all these `sendBlocks` methods :D So we have 2 options:

- either use std::vector<std::shared_ptr<DagBlock>>&&
- or const std::vector<std::shared_ptr<DagBlock>>& 

depends what is more appropriate

File: libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp [hunk 11]
```diff
@@ -46,23 +41,28 @@ void DagBlockPacketHandler::process(const PacketData &packet_data, const std::sh
     }
 
     if (auto status = checkDagBlockValidation(block); !status.first) {
-      // Ignore new block packets when pbft syncing
+      // Ignore new block packets when pbft syncing or dag syncing
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/dag_block_packet_handler.cpp:44]: either the comment is wrong or is_pbft_syncing() name is wrong here

File: libraries/core_libs/network/src/tarcap/packets_handlers/dag_sync_packet_handler.cpp [hunk 12]
```diff
@@ -55,12 +79,10 @@ void DagSyncPacketHandler::process(const PacketData& packet_data, const std::sha
     dag_blk_mgr_->insertBroadcastedBlock(std::move(block));
   }
 
-  if (missing_blks.size() > 0) {
-    requestDagBlocks(packet_data.from_node_id_, missing_blks, DagSyncRequestType::MissingHashes);
-  }
-  syncing_state_->set_dag_syncing(false);
+  peer->peer_dag_synced = true;
+  peer->peer_dag_syncing = false;
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/dag_sync_packet_handler.cpp:83]: so as dag_syncing is moved out from syncing_state, once we are pbft synced (but not yet dag synced), received dag blocks from other peers will probably fail because we are not dag synced yet, is that right and are we ok with it ?

Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/packets_handlers/dag_sync_packet_handler.cpp:83]: + didnt you changed testnet that we are sending dag blocks in batches of 100 ? In such case this would not work properly ?

**Review by JakubFornadel — APPROVED** [review]

File: libraries/core_libs/network/src/tarcap/packets_handlers/dag_sync_packet_handler.cpp (hunk 12 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/dag_sync_packet_handler.cpp:83]: The batch of 100 is removed from testnet as well. This was there only because I originally did not realize what was causing the slow processing. And yes we are ok with received dag blocks failing with tips/pivot missing until dag_synced is true, once dag_synced is true for a peer any missing tip/pivot should indicate a malicious node. Currently I disconnect such a node but we should have some blacklisting as well not to reconnect to it again

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 13]
```diff
@@ -414,6 +417,13 @@ void DagManager::addDagBlock(DagBlock const &blk, SharedTransactions &&trxs, boo
       trx_mgr_->saveTransactionsFromDagBlock(trxs);
       // Save the dag block
       db_->saveDagBlock(blk);
+
+      // TODO: This is an ugly temporary fix for testnet, a better solution is needed for dag block race condition
+      if (db_->getDagBlockPeriod(blk.getHash()) != nullptr) {
+        db_->removeDagBlock(blk.getHash());
+        LOG(log_er_) << "Block already in DB: " << blk.getHash();
+        return;
+      }
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/src/dag/dag.cpp:426]: I am not sure if you will fix it in a better way, or just leave it as this

**Review by gaoqi8649 — APPROVED** [review]

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 13 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:426]: I will fix this but I will need another PR for this since the solution is not simple
